File: 0377-combination-sum-iv/0377-combination-sum-iv.java
class Solution {
    Integer[] dp = new Integer[1001];
    public int combinationSum4(int[] nums, int target) {
        if(target == 0) return 1;
        if(target <  0) return 0;
        if(dp[target]!=null) return dp[target];
        int totalCombination = 0;
        for(int i=0;i<nums.length;i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int num = nums[i]; 
            totalCombination+=combinationSum4(nums, target-num);
        }

        return dp[target] = totalCombination;
    }
}

File: 0377-combination-sum-iv/0377-combination-sum-iv.cpp
class Solution {
public:
    int helper(vector<int>&nums, int target, vector<int>&dp)
    {
        if(target==0)
            return 1;
        if(dp[target]!=-1) return dp[target];
        dp[target]=0;
        for(auto&ell:nums)
        {
            if(ell<=target) dp[target]+=helper(nums,target-ell,dp);
        }
        return dp[target];
    }
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> dp(target+1, -1);
        return helper(nums, target, dp);
    }
};